New test for verifying one-sided MPI communication

nancycollins opened the pull request:
## Description:
<!--- Describe your changes -->
Add a new MPI one-sided test to verify MPI is working correctly.   Update other tests to use the standard 'system' block formatting, add a version of fixsystem here.  Add new test to Makefile and update it.  Update the README to fix directory names, contact info, add ftest_onesided info.

i also have some minor non-code changing updates to other files in this dir.  Will commit them separately later.

### Fixes issue
<!--- link to github issue(s) -->
Currently there is a user report of problems with one-sided communication.  This standalone test should verify that the one-sided MPI functions work.  (Even if MPI works, sometimes there are ways for one-sided to break.)

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ x] Documentation update

### Documentation changes needed?
<!-- Put an `x` in all the boxes that apply: -->
- [ x] My change requires a change to the documentation.
  - [x ] I have updated the documentation accordingly.

### Tests
Built and ran in this directory on cheyenne compute node (needed for MPI0 with intel and gfortran.

## Checklist for merging

- [ ] Updated changelog entry
- [ x] Documentation updated
- [ ] Version tag 

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ x] No dataset needed

nancycollins commented:
i have updated versions of the remaining tests in this dir that simply remove the older SVN lines.  they aren't worth a separate release, but i would like to update them at some point.  can they be part of this pull request or should they be a separate pull request?  sorry, i am unclear on how 'non-code-change-cleanup' things are being handled.

hkershaw-brown commented:
quick question on the fixsystem.  Can you use the fixsytem script already in DART?

nancycollins commented:
> quick question on the fixsystem. Can you use the fixsytem script already in DART?

great question.  i looked at it and the differences with in the one in the utilities dir are:

1) it has the mpi_utilities_mod.f90 and null_mpi_utilities_mod.f90 filenames hardcoded in it.
2) it fails if it can't find the key strings in the files.

i looked at trying to make it take filenames as an argument but it already takes the compiler name as an optional arg.  the compiler name could be changed to be required and not optional, and then followed by the filenames to process, with the mpi modules as a default.  but those 2 files must have the strings in them and i'd still want it to fail if they aren't found.  (that might mean they've been corrupted, or edited by hand and no longer have the key strings that sed needs to work.)  only some of the files in the test directory have a call to system() and so some of the test files don't need to be changed.  i could list the exact test files that do need editing, i suppose, in the call in the Makefile to fixsystem if i could work out how to supply optional filenames in addition to a compiler type.

i don't like replicating the functionality but wasn't sure i wanted to mess with the utilities directory version.  that will need a lot more testing.  you can convince me that's a lesser evil compared to replicating the file.

nancycollins commented:
i've updated the utilities version of fixsystem to work with the test files, and removed the copy of fixsystem.  this still needs to be tested with NAG because i think the original fixsystem logic was wrong.  waiting to see if i can get access to one before it's ready to merge.

nancycollins commented:
ok.  i've tested this with the NAG compiler on hobart.cgd.ucar.edu.   i've had to update the mkmf.template file to get DART to build ok.  i've removed the go/stop mpi tests because no one uses async4 anymore.  they're in the history if they need to be revived.  i've updated the ftest_mpi test to have the same conditional blocks as the current mpi modules, and it's now using the utilities directory version of fixsystem, so i've removed the copy here.

nancycollins commented:
i haven't done anything to the version or changelog files.

hkershaw-brown commented:
Hey Nancy, 

Can you add what changes are in this pull request to description 
I think it is this:
* new mpi one-sided test
* bug-fix(?) for fixsystem for the nag compiler
* removed async 4 tests
*  added flag to mkmf.template.nag.linux (is this just for the tests or for general DART compiling?)
* portability-fix(?) for strings.h (-> string.h)
* bug-fix: use `stop` instead of `call exit`

A couple of questions/paranoia:
1. . There are still references to async and asyn4 so I am a bit wary of removing the tests, e.g.
`models/lorenz_96/shell_scripts/run_filter_async4.csh`
https://docs.dart.ucar.edu/en/latest/guide/filter_async_modes.html?highlight=async
What do you think?
2.  Using -mismatch with mpi is the f77 vs mpi08 problem #158 #261.  It seems a bit strange to use the nag compiler, but downgrade the consistency checking of procedure argument lists.  Do you think nag compile updates should be with an update to use mpif08?

nancycollins commented:
i updated the description, but the changelog and version still need work.   to answer your questions:

1) i think we should eventually remove all references to async4.  while it worked, it was fragile and overly complex.  i think if we need to revive that behavior there may be newer MPI facilities that allow one MPI program to spawn another, or there may be other solutions that aren't so cryptic.  i vote async4 off the island.

2) i looked that those issues with MPI and NAG and yes, it's exactly the same problem.  we have a subroutine in the mpi_utilities_mod.f90 file that first calls MPI_Ssend() with a real array, and then calls MPI_Ssend() with an integer array.  that's perfectly fine for MPI - there's an argument to tell MPI what the array type is.  but the compiler freaks out if they're both used in the same subroutine.  i guess for now i'd leave -mismatch in the nag mkmf.template, because it still prints out warnings if it sees mismatching args but it doesn't error out.    this is needed for building dart, not just the tests.  if derecho is going to have NAG installed then we probably will have to address this more completely in the future, but for now i think it's an ok intermediate step.  i will add a comment in the mkmf.template.nag.linux file about this so there's a reminder of why that option is used.  if/when 'use mpi_f08' is widely available, we can start using that in our code instead of just 'use mpi'.

hkershaw-brown commented:
> 
> 1. i think we should eventually remove all references to async4.  while it worked, it was fragile and overly complex.  i think if we need to revive that behavior there may be newer MPI facilities that allow one MPI program to spawn another, or there may be other solutions that aren't so cryptic.  i vote async4 off the island.

So,  leave the tests in and remove asyn4 at a later date?

nancycollins commented:
the async4 tests, if they stay in, will need to be updated with the same conditional code that i added to ftest_mpi.f90.  it's a copy of what's in the real mpi_utilities_mod.f90.  these tests are supposed to be standalone (although ftest_sendrecv isn't - i'm not sure why and i don't remember the history there) so i can't re-use the utilities module code.  i took them out rather than update them.  if you really think they need to stay, let me know and i'll go back and add the conditional sections to them.

originally we didn't have the 'fixsystem' script, so people had to edit mpi_utilities_mod by hand to make it work.  when i added support for the NAG compiler there were a couple other conditional sections added that the script took care of.  but the tests were never updated to use the script, so to make them work someone would have had to edit them by hand if they used anything other than the gfortran compiler.  i've made them look more like how the dart code works so they should be easier for someone to run, but are still standalone in case they need to run a small program, or send their system support a small program as an example of something not working.

hkershaw-brown commented:
oh ok, fair enough, let's get rid of the async4 tests.

nancycollins commented:
the regression test for fixsystem is to build L96 with various compilers.  it should build.  

the regression test for the programs in the developer_tests/mpi_utilities/tests directory is to 'make everything' with various compilers.  they should build.

nancycollins commented:
i've finished messing around with the new ftest_onesided.f90 source.  i think i've cleaned it up and that the logic is simpler to follow, the subroutines are cleaner, and the comments are less wrong.   i think this code is ready for review.  i'll be gone for the next 4 days starting tomorrow, so no rush on this.
